src/Core/SymbolUtility.cs

```diff
@@ -623,53 +623,57 @@ public static ulong GetEnumValueAsUInt64(object value, INamedTypeSymbol enumType
+    /// </remarks>
+    /// <param name="semanticModel">Used to determine whether a candidate <c>GetAwaiter</c> method is accessible at the given <paramref name="position"/>.</param>
+    /// <param name="position">The position of the token at which the <paramref name="symbol"/> is used.</param>
+    /// <returns>A <see cref="bool"/> indicating whether the symbol is an awaitable type or a method that returns one.</returns>
+    public static bool IsAwaitable(this ISymbol? symbol, SemanticModel semanticModel, int position)
     {
-        INamedTypeSymbol? namedTypeSymbol = GetPossiblyAwaitableType(typeSymbol);
+        ITypeSymbol? typeSymbol = (symbol as ITypeSymbol) ?? (symbol as IMethodSymbol)?.ReturnType;
 
-        if (namedTypeSymbol is null)
+        if (typeSymbol is null or { SpecialType: SpecialType.System_Void })
             return false;
 
-        INamedTypeSymbol originalDefinition = namedTypeSymbol.OriginalDefinition;
-        TypeKind typeKind = originalDefinition.TypeKind;
+        // this is the same check as in Roslyn, reimplemented due to it being internal
+        // https://github.com/dotnet/roslyn/blob/a182892bf997a457cfcdbece5352e1a139eb2a12/src/Compilers/CSharp/Portable/Binder/Binder_Await.cs#L281-L289
 
-        if (typeKind == TypeKind.Struct
-            && originalDefinition.ContainingNamespace.HasMetadataName(MetadataNames.System_Threading_Tasks))
-        {
-            switch (originalDefinition.MetadataName)
-            {
-                case "ValueTask":
-                case "ValueTask`1":
-                    return true;
-            }
-        }
+        IMethodSymbol? getAwaiter = semanticModel.LookupSymbols(position, typeSymbol, WellKnownMemberNames.GetAwaiter, includeReducedExtensionMethods: true)
+            .OfType<IMethodSymbol>()
+            .FirstOrDefault();
 
-        if (typeKind == TypeKind.Class
-            && namedTypeSymbol.EqualsOrInheritsFrom(MetadataNames.System_Threading_Tasks_Task))
-        {
-            return true;
-        }
+        if (getAwaiter is not { Parameters.Length: 0 })
+            return false;
 
-        if (shouldCheckWindowsRuntimeTypes)
-        {
-            if (typeKind == TypeKind.Interface
-                && originalDefinition.ContainingNamespace.HasMetadataName(MetadataNames.WinRT.Windows_Foundation))
-            {
-                switch (originalDefinition.MetadataName)
-                {
-                    case "IAsyncAction":
-                    case "IAsyncActionWithProgress`1":
-                    case "IAsyncOperation`1":
-                    case "IAsyncOperationWithProgress`2":
-                        return true;
-                }
-            }
+        var awaiterTypeDefinition = getAwaiter.ReturnType.OriginalDefinition as INamedTypeSymbol;
+        if (awaiterTypeDefinition is null or { SpecialType: SpecialType.System_Void })
```

Comment:
Same as previous comment: Could you invert the condition and use `SpecialType.None`?

Reply:
This case does apply - good catch!